### PR TITLE
[3.13] Add the Python 3.13 abidump, and enable the ABI check.

### PR DESCRIPTION
**Pull request by Yhg1s**
Now that we're in release-candidate mode, add the Python 3.13 ABI dump file, and enable the ABI check workflow. (Copied from the 3.12 workflow with s/ubuntu-22.04/ubuntu-latest/ since that seems to be the fashion now.)

----
📚 Documentation preview 📚: https://cpython-previews--122583.org.readthedocs.build/

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -40,6 +40,49 @@ jobs:
     if: fromJSON(needs.check_source.outputs.run-docs)
     uses: ./.github/workflows/reusable-docs.yml
 
+  check_abi:
+    name: 'Check if the ABI has changed'
+    runs-on: ubuntu-latest
+    needs: check_source
+    if: needs.check_source.outputs.run_tests == 'true'
+    steps:
+      - uses: actions/checkout@v3
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
      - uses: actions/checkout@v4
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -40,6 +40,49 @@ jobs:
...
+      - uses: actions/setup-python@v4
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
      - uses: actions/setup-python@v5
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -40,6 +40,49 @@ jobs:
...
+      - name: Check for changes in the ABI
+        id: check
+        run: |
+          if ! make check-abidump; then
+            echo "Generated ABI file is not up to date."
+            echo "Please add the release manager of this branch as a reviewer of this PR."
+            echo ""
+            echo "The up to date ABI file should be attached to this build as an artifact."
+            echo ""
+            echo "To learn more about this check: https://devguide.python.org/setup/#regenerate-the-abi-dump"
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
            echo "To learn more about this check: https://devguide.python.org/getting-started/setup-building/index.html#regenerate-the-abi-dump"
```

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -40,6 +40,49 @@ jobs:
...
+            echo ""
+            exit 1
+          fi
+      - name: Generate updated ABI files
+        if: ${{ failure() && steps.check.conclusion == 'failure' }}
+        run: |
+          make regen-abidump
+      - uses: actions/upload-artifact@v3
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
      - uses: actions/upload-artifact@v4
```

**Comment by Yhg1s** [comment]
Should those changes be made to 3.12 as well?

**Comment by hugovk** [comment]
The actions have already been updated in 3.12 (https://github.com/python/cpython/pull/116948).

The URL update doesn't seem important enough on its own, there's a redirect in place.

**Review by hugovk — APPROVED** [review]

**Review by encukou — APPROVED** [review]
With `ubuntu-latest` we might get spurious ABI complaints when Ubuntu updates. But it's just one of the ways that update *might* bite us, and this one should be easy to solve/revisit.

**Pull request merged**